server/routes/utils/reportHelper.ts

```diff
@@ -14,89 +14,136 @@
-      fullPage: true,
-    });
-
-    //TODO: Add header and footer, phase 2
-
-    await browser.close();
-    return { timeCreated, dataUrl: buffer.toString('base64'), fileName };
+    return renderReport(
+      'image/png',
+      url,
+      source,
+      itemName,
+      windowWidth,
+      windowHeight,
+      header,
+      footer
+    );
   } catch (error) {
     throw error;
   }
 };
 
 export const generatePDF = async (
   url: string,
+  source: string,
   itemName: string,
   windowWidth: number,
-  windowHeight: number
+  windowHeight: number,
+  header: string,
+  footer: string
 ): Promise<{ timeCreated: string; dataUrl: string; fileName: string }> => {
   try {
-    const browser = await puppeteer.launch({
-      headless: true,
-    });
-    const page = await browser.newPage();
+    return renderReport(
+      'application/pdf',
+      url,
+      source,
+      itemName,
+      windowWidth,
+      windowHeight,
+      header,
+      footer
+    );
+  } catch (error) {
+    throw error;
+  }
+};
 
-    await page.setDefaultNavigationTimeout(0);
-    await page.setViewport({
-      width: windowWidth,
-      height: windowHeight,
-    });
+async function renderReport(
+  format: string,
+  url: string,
+  source: string,
+  itemName: string,
+  windowWidth: number,
+  windowHeight: number,
+  header: string,
+  footer: string
+) {
+  if (source === 'Dashboard') {
+    windowWidth = 1440;
+    windowHeight = 2560;
+  } else if (source === 'Visualization') {
+    windowWidth = 900;
+    windowHeight = 1300;
+  }
 
-    await page.goto(url, { waitUntil: 'networkidle0' });
+  const browser = await puppeteer.launch({
+    headless: true,
+  });
+  const page = await browser.newPage();
 
-    const timeCreated = new Date().toISOString();
-    const fileName = getFileName(itemName, timeCreated) + '.' + FORMAT.pdf;
-    // The scrollHeight value is equal to the minimum height the element would require in order to fit
-    // all the content in the viewport without using a vertical scrollbar
-    const scrollHeight = await page.evaluate(
-      () => document.documentElement.scrollHeight
-    );
+  await page.setDefaultNavigationTimeout(0);
+  await page.setViewport({
+    width: windowWidth,
+    height: windowHeight,
+  });
+
+  await page.goto(url, { waitUntil: 'networkidle0' });
+  let element;
+
+  if (source === 'Dashboard') {
+    await page.waitForSelector('#dashboardViewport');
+    element = await page.$('#dashboardViewport');
+  } else if (source === 'Visualization') {
+    await page.waitForSelector('.visChart');
+    element = await page.$('.visChart');
+  }
+
+  const screenshot = await element.screenshot({ fullPage: false });
+  const timeCreated = new Date().toISOString();
 
-    const buffer = await page.pdf({
+  const scrollHeight = await page.evaluate(
+    () => document.documentElement.scrollHeight
+  );
+
+  await page.setContent(`
+  <!DOCTYPE html>
+  <html>
+    <div>
+    <h1>${header}</h1>
```

Comment:
Just curious if it possible to provide the header directly instead of taking screenshot and providing screenshot? may be use frames? May be it will help copy-able text in pdf format?

Reply:
As far as I know, to provide the header we have to set the page content and add it. I will investigate further though